tutorials/08-voting/contracts/Voting.cdc

```diff
@@ -0,0 +1,132 @@
+/*
+*
+*   In this example, we want to create a simple voting contract
+*   where a polling place issues ballots to addresses.
+*
+*   The run a vote, the Admin deploys the smart contract,
+*   then initializes the proposals.
+*   The array of proposals cannot be modified after it has been initialized.
+*
+*   Users can create ballots and vote only with their GovernanceToken balance prior to when
+*   proposal was created.
+*
+*   Every user with a ballot is allowed to approve their chosen proposals.
+*   A user can choose their votes and cast them
+*   with the tx_05_SelectAndCastVotes.cdc transaction.
+*
+*/
+
+import GovernanceToken from "./GovernanceToken.cdc"
+
+pub contract Voting {
+
+    // list of proposals to be approved
+    pub var proposals: [ProposalData]
+
+    // paths
+    pub let adminStoragePath: StoragePath
+    pub let ballotStoragePath: StoragePath
+    pub let ballotPublicPath: PublicPath
+
+    pub struct ProposalData {
+        pub let name: String
+        pub let checkpoint: UInt16
+        pub(set) var votes: UFix64
+        pub(set) var voters: {UInt64: Bool}
+
+        init(name: String) {
+            self.name = name
+            self.checkpoint = GovernanceToken.checkpointCounter
+            self.votes = 0.0
+            self.voters = {}
+        }
+    }
+
+    pub resource interface Votable {
+        pub vaultId: UInt64
+        pub votingWeightDataSnapshot: [GovernanceToken.VotingWeightData]
+
+        pub fun vote(proposalId: Int){
+            pre {
+                Voting.proposals[proposalId] != nil: "Cannot vote for a proposal that doesn't exist"
+                Voting.proposals[proposalId].voters[self.vaultId] == nil: "Cannot cast vote again using same Governance Token Vault"
+                self.votingWeightDataSnapshot != nil && self.votingWeightDataSnapshot.length > 0: "Can only vote if balance exists"
+                self.votingWeightDataSnapshot[0].checkpoint < Voting.proposals[proposalId].checkpoint: "Can only vote if balance was recorded before proposal was created"
+            }
+        }
+    }
+
+    // This is the resource that is issued to users.
+    // When a user gets a Ballot resource, they call the `vote` function
+    // to include their votes
+    pub resource Ballot: Votable {
+        // id of GovernanceToken Vault
+        pub let vaultId: UInt64
+        // array of GovernanceToken Vault's votingWeightDataSnapshot
+        pub let votingWeightDataSnapshot: [GovernanceToken.VotingWeightData]
```

Comment:
Might be better to make this a dictionary keyed by the checkpoint number to make it easier to lookup

Reply:
Not using a checkpoint anymore but the block timestamp, so unfortunately, dictionary lookup won't work.